### PR TITLE
hep: populate documents and figures from FFT

### PR DESCRIPTION
**Pull request by zzacharo**
Deletes `_fft` field populated from FFT and introduces
`documents` and `figures` instead.

Signed-off-by: Zacharias Zacharodimos <zacharias.zacharodimos@cern.ch>

### CODE REVIEW

File: inspire_dojson/hep/rules/bdFFT.py [hunk 1]
```diff
@@ -24,60 +24,45 @@
...
-        's': _get_s(value),
-        't': value.get('type'),
-        'v': value.get('version'),
-        'z': value.get('status'),
-    }
+def documents(self, key, value):
+    """Populate the ``documents`` key.
+
+        Also populates the ``figures`` key through side effects.
+        """
```
Comment (jacquerie) [inline inspire_dojson/hep/rules/bdFFT.py]: Weird indentation?

File: inspire_dojson/hep/rules/bdFFT.py [hunk 2]
```diff
@@ -24,60 +24,45 @@
...
+    def _is_hidden(value):
+        return 'HIDDEN' in value or None
+
+    def _get_caption(value):
+        import re
```
Comment (jacquerie) [inline inspire_dojson/hep/rules/bdFFT.py]: There's no need for internal imports. Just put it on the top!

**Comment by zzacharo** [comment]
@michamos @jacquerie can you have a quick look on this?

File: inspire_dojson/hep/rules/bdFFT.py [hunk 3]
```diff
@@ -24,60 +24,45 @@
...
+        return ''.join(re.compile('^\d{5}\s+').split(value)[1:]) or None
+
+    _type = value.get('t', None)
```
Comment (jacquerie) [inline inspire_dojson/hep/rules/bdFFT.py]: `dict.get` already returns `None` if `'t'` is not there. Also, nit, you prepend `_` when you want to say that something is private, while you append it if you want to distinguish it from builtins. So here it should be `type_`.

File: inspire_dojson/hep/rules/bdFFT.py [hunk 4]
```diff
@@ -24,60 +24,45 @@
...
+
+    if _type == 'Plot':
+        _figures = self.get('figures', [])
+        _figures.append({
+            'key': value.get('n') + value.get('f', ''),
+            'caption': _get_caption(value.get('d')),
+            'url': afs_url(value.get('a'))
```
Comment (jacquerie) [inline inspire_dojson/hep/rules/bdFFT.py]: I know I used to do this, but I now prefer to pass the entire `value` to the helper, so that it (potentially) becomes reusable.

File: inspire_dojson/hep/rules/bdFFT.py [hunk 5]
```diff
@@ -24,60 +24,45 @@
...
+        return 'HIDDEN' in value or None
+
+    def _get_caption(value):
+        import re
+        return ''.join(re.compile('^\d{5}\s+').split(value)[1:]) or None
+
+    _type = value.get('t', None)
+
+    if _type == 'Plot':
+        _figures = self.get('figures', [])
```
Comment (jacquerie) [inline inspire_dojson/hep/rules/bdFFT.py]: Do this at the top. It's clearer. Also you don't need the `_` !

File: tests/test_hep_bdFFT.py [hunk 6]
```diff
@@ -49,106 +49,46 @@ def test_fft_from_FFT():
...
-        '  <subfield code="t">APS</subfield>'
-        '</datafield>'
-    )  # record/1094156
+    assert validate(result['documents'], subschema) is None
+    assert expected == result['documents']
 
-    assert '_fft' not in hep.do(create_record(snippet))
 
-
-def test_fft_from_FFT_ignores_context():
```
Comment (jacquerie) [inline tests/test_hep_bdFFT.py]: Adapt the test to the new code. There was a point to this test, if you delete it you lose the knowledge that is stored in it!

File: inspire_dojson/hep/rules/bdFFT.py [hunk 7]
```diff
@@ -24,60 +24,45 @@
...
-        'v': value.get('version'),
-        'z': value.get('status'),
-    }
+def documents(self, key, value):
+    """Populate the ``documents`` key.
+
+        Also populates the ``figures`` key through side effects.
+        """
+    def _is_hidden(value):
+        return 'HIDDEN' in value or None
```
Comment (jacquerie) [inline inspire_dojson/hep/rules/bdFFT.py]: You'll probably want to `.upper()` this value to avoid false negatives (`'hidden'`). Don't trust that the data is correct!

File: inspire_dojson/utils/__init__.py [hunk 8]
```diff
@@ -98,6 +98,19 @@ def absolute_url(relative_url):
...
 
 
+def afs_url(eos_path):
+    """Returns the afs absolute path from eos path.
+
+    The base URL is taken from the Flask app config if present, otherwise it
+    falls back to ``http://inspirehep.net``.
+    """
+    default_afs_path = '/afs/cern.ch/project/inspire/PROD/'
+    file_path = current_app.config.get('FILE_SERVER_PATH', default_afs_path)
```
Comment (jacquerie) [inline inspire_dojson/utils/__init__.py]: Why do we need this configurability? We already know in advance which path we're going to get.

We should remove, not reinforce, the dependency of `inspire-dojson` from Flask.

File: inspire_dojson/utils/__init__.py [hunk 9]
```diff
@@ -98,6 +98,19 @@ def absolute_url(relative_url):
...
+    if re.match('^/opt/cds-invenio', eos_path):
```
Comment (michamos) [inline inspire_dojson/utils/__init__.py]: `if eos_path.startswith('/opt/cds-invenio'):` is clearer, but the proper way to do this kind of things is with `os.path` (see https://stackoverflow.com/a/27258779).

File: inspire_dojson/utils/__init__.py [hunk 10]
```diff
@@ -98,6 +98,19 @@ def absolute_url(relative_url):
     return urllib.parse.urljoin(server, relative_url)
 
 
+def afs_url(eos_path):
```
Comment (michamos) [inline inspire_dojson/utils/__init__.py]: you are not passing an EOS path. So you shouldn't name your argument like that.

File: tests/test_hep_bdFFT.py (hunk 6 above)
Comment (michamos) [reply tests/test_hep_bdFFT.py]: the point is that you should ignore FFT tags whose `f` subfield ends in context.

File: tests/test_hep_bdFFT.py [hunk 11]
```diff
@@ -49,106 +49,46 @@ def test_fft_from_FFT():
...
+        '  <subfield code="f">.png</subfield>'
+        '  <subfield code="n">EA-vs-n</subfield>'
+        '  <subfield code="r"/>'
+        '  <subfield code="s">2016-04-01 15:14:38</subfield>'
+        '  <subfield code="t">Plot</subfield>'
+        '  <subfield code="v">1</subfield>'
+        '  <subfield code="z"/>'
         '</datafield>'
-    )  # DESY harvest
+    )
```
Comment (michamos) [inline tests/test_hep_bdFFT.py]: you should write the recid in a comment, as is done for the other snippets.

**Comment by zzacharo** [comment]
@jacquerie I updated the code. Check it again!

File: inspire_dojson/hep/rules/bdFFT.py [hunk 12]
```diff
@@ -24,60 +24,45 @@
...
+
+    def _get_caption(value):
+        import re
+        return ''.join(re.compile('^\d{5}\s+').split(value)[1:]) or None
```
Comment (michamos) [inline inspire_dojson/hep/rules/bdFFT.py]: the number is used to order the figures and you should make sure that the order of figures in the final `figures` array is the same as the one given by this numerical ordering (please also add a test for that). You probably need to remove the `for_each_value` decorator and iterate over the values yourself to be able to control the ordering (you can look at `figures2marc` for an example).

**Comment by michamos** [comment]
@zzacharo please have a look at my (now hidden) comments.

File: tests/test_hep_bdFFT.py [hunk 13]
```diff
@@ -115,42 +111,6 @@ def test_fft_from_FFT_ignores_context():
     assert '_fft' not in result
 
 
-def test_fft_from_FFT_does_not_require_s():
```
Comment (jacquerie) [inline tests/test_hep_bdFFT.py]: I think you should preserve this test too. It's true that we no longer use `FFT__s`, but it was here because it documented a bug fix from which we should not regress.

Comment (jacquerie) [reply tests/test_hep_bdFFT.py]: Same goes for `test_fft_from_FFT_percent_percent`: AFAIK it no longer happens in Legacy, but we should keep the test, otherwise people see the regex `FFT[^%][^%]` and have no idea why it's there.

**Comment by michamos** [comment]
@jacquerie `for_each_value` is not doing what we think it does. We need to discuss this.

**Comment by zzacharo** [comment]
@michamos @jacquerie review this commit https://github.com/inspirehep/inspire-dojson/pull/105/commits/fc5049161fa3f1815b823abea5f692c770eb5e54 .

File: inspire_dojson/hep/rules/bdFFT.py [hunk 14]
```diff
@@ -45,32 +46,49 @@ def _is_hidden(value):
...
+
+    def _get_key(value):
+        fname = value.get('n', str(uuid.uuid4()))
+        extension = value.get('f', '')
+
+        if fname.endswith(extension):
+            return fname
+        return fname + extension
+
+    ordered_pattern = re.compile('^(\d{5})\s*')
```
Comment (michamos) [inline inspire_dojson/hep/rules/bdFFT.py]: why do you need two patterns? it would be simpler to modify `_get_caption` to `_get_index_and_caption` that would return both the `index` (e.g. `00005` or `None`) and the `caption`.

File: inspire_dojson/hep/rules/bdFFT.py [hunk 15]
```diff
@@ -45,32 +46,49 @@ def _is_hidden(value):
         return 'HIDDEN' in [val.upper() for val in value] or None
 
     def _get_caption(value):
-        return ''.join(re.compile('^\d{5}\s+').split(value)[1:]) or None
+        return ''.join(re.compile('^\d{5}\s*').split(value)[1:]) or None
+
+    def _get_key(value):
+        fname = value.get('n', str(uuid.uuid4()))
```
Comment (michamos) [inline inspire_dojson/hep/rules/bdFFT.py]: `uuid`s are not nice.

File: inspire_dojson/hep/rules/bdFFT.py [hunk 16]
```diff
@@ -45,32 +46,49 @@ def _is_hidden(value):
...
-        document = {
-            'description': value.get('d'),
-            'key': value.get('n') + value.get('f', ''),
-            'hidden': _is_hidden(force_list(value.get('o'))),
-            'url': afs_url(value)
-        }
+    format_ = value.get('f', '')
+    ordnum_match = ordered_pattern.match(value.get('d'))
+    if format_ in figures_extensions:
+        if ordnum_match:
```
Comment (michamos) [inline inspire_dojson/hep/rules/bdFFT.py]: if you implement my previous suggestion, you don't need this test: the function would return `None` if there is no `index`, and `None` values are cleaned up at the end by dojson.

File: inspire_dojson/hep/rules/bdFFT.py [hunk 17]
```diff
@@ -45,32 +46,49 @@ def _is_hidden(value):
...
-    elif type_ == 'Main':
-        document = {
-            'description': value.get('d'),
-            'key': value.get('n') + value.get('f', ''),
-            'hidden': _is_hidden(force_list(value.get('o'))),
-            'url': afs_url(value)
-        }
+    format_ = value.get('f', '')
+    ordnum_match = ordered_pattern.match(value.get('d'))
+    if format_ in figures_extensions:
```
Comment (michamos) [inline inspire_dojson/hep/rules/bdFFT.py]: for clarity, please create an `_is_figure` function that does this test.

File: inspire_dojson/hep/rules/bdFFT.py [hunk 18]
```diff
@@ -45,32 +46,49 @@ def _is_hidden(value):
...
+        else:
+            figures.append({
+                'key': _get_key(value),
+                'caption': _get_caption(value.get('d')),
+                'url': afs_url(value)
+            })
 
     self['figures'] = figures
-    return document
+    return {
```
Comment (michamos) [inline inspire_dojson/hep/rules/bdFFT.py]: this looks wrong: if it's a figure, it also gets executed. You should assert in the tests that when you populate `figures`, `'documents' not in record` and vice-versa.

File: tests/test_hep_bdFFT.py [hunk 19]
```diff
@@ -22,6 +22,8 @@
 
 from __future__ import absolute_import, division, print_function
 
+import re
```
Comment (michamos) [inline tests/test_hep_bdFFT.py]: Unused import.

File: tests/test_hep_bdFFT.py [hunk 20]
```diff
@@ -108,7 +172,57 @@ def test_documents_from_FFT_ignores_context():
...
+    expected = [
+        {
+            'description': 'Fulltext',
+            'url': 'http://www.mdpi.com/2218-1997/3/1/24/pdf'
+        }
+    ]
+    result = hep.do(create_record(snippet))
+
+    assert validate(result['documents'], subschema) is None
+    expected[0]['key'] = result['documents'][0]['key']
```
Comment (michamos) [inline tests/test_hep_bdFFT.py]: you shouldn't cheat in tests like that. If you want reproducibility, you need to mock the random generator you are using.

File: tests/test_hep_bdFFT.py [hunk 21]
```diff
@@ -108,7 +172,57 @@ def test_documents_from_FFT_ignores_context():
...
+def test_fft_from_FFT_percent_percent():
+    snippet = (
+        '<datafield tag="FFT" ind1="%" ind2="%">'
+        '  <subfield code="a">/opt/cds-invenio/var/tmp-shared/apsharvest_unzip_5dGfY5/articlebag-10-1103-PhysRevD-87-083514-apsxml/data/PhysRevD.87.083514/fulltext.xml</subfield>'
+        '  <subfield code="o">HIDDEN</subfield>'
+        '  <subfield code="t">APS</subfield>'
+        '</datafield>'
+    )  # record/1094156
+
+    assert 'documents' not in hep.do(create_record(snippet))
```
Comment (michamos) [inline tests/test_hep_bdFFT.py]: if you define `result = hep.do(create_record(snippet))`, you can reuse it instead of computing it twice.

File: inspire_dojson/hep/model.py [hunk 22]
```diff
@@ -83,12 +83,37 @@ def convert_curated(record, blob):
...
 
 
+def ensure_ordered_figures(record, blob):
+    ordered_figures_dict = {}
+    unordered_figures_list = []
+
+    for figure in record.get('figures', []):
+        order = figure.get('order')
+        if order:
+            figure.pop('order')
```
Comment (michamos) [inline inspire_dojson/hep/model.py]: you can simplify this to 
```python
order = figure.pop('order', None):
if order:
    ordered_figures_dict[order] = figure
else:
    unordered_figures_list.append(figure)
```

File: inspire_dojson/hep/model.py [hunk 23]
```diff
@@ -83,12 +83,37 @@ def convert_curated(record, blob):
...
+            ordered_figures_dict[order] = figure
+        else:
+            unordered_figures_list.append(figure)
+
+    if ordered_figures_dict:
```
Comment (michamos) [inline inspire_dojson/hep/model.py]: all of this `if ... else ...` ca be simplified to:
```python
record['figures'] = [value for key, value in sorted(six.iteritems(ordered_figures_dict))]
record['figures'].extend(unordered_figures_list)
```
So you skip the test, and in the case you have both ordered and unordered figures (that probably doesn't happen, but better safe than sorry), you don't throw away part of them.

File: inspire_dojson/hep/model.py [hunk 24]
```diff
@@ -83,12 +83,37 @@ def convert_curated(record, blob):
...
+        sorted_keys = ordered_figures_dict.keys()
+        sorted_keys.sort()
+
+        ordered_figures_list = [ordered_figures_dict[key] for key in sorted_keys]
+        ordered_figures_list.append(unordered_figures_list)
+        record['figures'] = ordered_figures_list
+    else:
+        record['figures'] = unordered_figures_list
```
Comment (michamos) [inline inspire_dojson/hep/model.py]: and here you just do `record['figures'].extend(unordered_figures_list)`. So you skip the test, and in the case you have both ordered and unordered figures (that probably doesn't happen, but better safe than sorry), you don't throw away part of them.

**Comment by zzacharo** [comment]
@michamos @jacquerie review https://github.com/inspirehep/inspire-dojson/pull/105/commits/433a16f140b4864a7cf7999d9318b18f993c8f21 .

File: inspire_dojson/hep/model.py [hunk 25]
```diff
@@ -83,12 +86,46 @@ def convert_curated(record, blob):
...
+        else:
+            unordered_figures_list.append(figure)
+
+    record['figures'] = [value for key, value in sorted(six.iteritems(ordered_figures_dict))]
+    record['figures'].extend(unordered_figures_list)
+    return record
+
+
+def ensure_unique_documents_and_figures(record, blob):
+    def dupliactes(elements):
```
Comment (michamos) [inline inspire_dojson/hep/model.py]: typo: `duplicates`

File: inspire_dojson/hep/model.py [hunk 26]
```diff
@@ -83,12 +86,46 @@ def convert_curated(record, blob):
...
+        duplicate_keys_list = []
+        for index, element in enumerate(elements):
+            if element:
```
Comment (michamos) [inline inspire_dojson/hep/model.py]: why is this `if` needed?

Comment (zzacharo) [reply inspire_dojson/hep/model.py]: Because we clean the record after we have some times list [None] and then it crashes. I wanted to keep the cleaning step for last otherwise I can move the check after cleaning. What do you think?

Comment (michamos) [reply inspire_dojson/hep/model.py]: OK then.

**Comment by zzacharo** [comment]
@jacquerie Can you check also and if is it ok for you too I squash the commits and merge it?

**Comment by jacquerie** [comment]
> @jacquerie Can you check also and if is it ok for you too I squash the commits and merge it?

I enjoy squashing commits too much to miss the opportunity of doing it : )
